planner/codec/codec.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codec
+
+import (
+	"bytes"
+	"compress/zlib"
+	"encoding/base64"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/errors"
+)
+
+const (
+	// TreeBody indicates the current operator sub-tree is not finished, still
+	// has child operators to be attached on.
+	TreeBody = '│'
+	// TreeMiddleNode indicates this operator is not the last child of the
+	// current sub-tree rooted by its parent.
+	TreeMiddleNode = '├'
+	// TreeLastNode indicates this operator is the last child of the current
+	// sub-tree rooted by its parent.
+	TreeLastNode = '└'
+	// TreeGap is used to represent the gap between the branches of the tree.
+	TreeGap = ' '
+	// TreeNodeIdentifier is used to replace the TreeGap once we need to attach
+	// a node to a sub-tree.
+	TreeNodeIdentifier = '─'
+)
+
+const (
+	rootTaskType = "0"
+	copTaskType  = "1"
+)
+
+const (
+	lineBreaker    = '\n'
+	lineBreakerStr = "\n"
+	separator      = '\t'
+	separatorStr   = "\t"
+)
+
+var decoderPool = sync.Pool{
+	New: func() interface{} {
+		return &planDecoder{}
+	},
+}
+
+// DecodePlan use to decode the string to plan tree.
+func DecodePlan(planString string) (string, error) {
+	if len(planString) == 0 {
+		return "", nil
+	}
+	pd := decoderPool.Get().(*planDecoder)
+	defer decoderPool.Put(pd)
+	pd.buf.Reset()
+	pd.decompressBuf.Reset()
+	return pd.decode(planString)
+}
+
+type planDecoder struct {
+	buf           bytes.Buffer
+	depths        []int
+	indents       [][]rune
+	planInfos     []*planInfo
+	decompressBuf bytes.Buffer
+}
+
+type planInfo struct {
+	depth  int
+	fields []string
+}
+
+func (pd *planDecoder) decode(planString string) (string, error) {
+	str, err := decompress(planString, &pd.decompressBuf)
+	if err != nil {
+		return "", err
+	}
+
+	nodes := strings.Split(str, lineBreakerStr)
+	if len(pd.depths) < len(nodes) {
+		pd.depths = make([]int, 0, len(nodes))
```

Comment:
Why not just let golang handle the reallocate?

Reply:
For pre-allocated memory.